### PR TITLE
ASoC: intel: sof_sdw: check sdw version when matching codec

### PR DESCRIPTION
**Pull request by bardliao**
Some codecs with the same part id but different soundwire version
have different configurations. So we have to separate them in
codec_info_list[].

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

### CODE REVIEW

File: sound/soc/intel/boards/sof_sdw_common.h [hunk 1]
```diff
@@ -52,6 +52,7 @@ enum {
 
 struct sof_sdw_codec_info {
 	const int id;
```
Comment (RanderWang) [inline sound/soc/intel/boards/sof_sdw_common.h]: how about change id to p_id ? it is confusing.

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw_common.h]: we can use 100 characters so let's use part_id and vendor_id. No need to use any encoding.

File: sound/soc/intel/boards/sof_sdw.c [hunk 2]
```diff
@@ -253,12 +253,21 @@ static struct sof_sdw_codec_info codec_info_list[] = {
...
+	part_id = SDW_PART_ID(adr);
+	sdw_version = SDW_VERSION(adr);
 	for (i = 0; i < ARRAY_SIZE(codec_info_list); i++)
-		if (part_id == codec_info_list[i].id)
+		/*
+		 * A codec info is for all sdw version with the part id if
+		 * v_id is not specified in the codec info.
+		 */
+		if (part_id == codec_info_list[i].id &&
+		    (!codec_info_list[i].v_id ||
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: do we need this test? The version start at one, it's never zero for all SoundWire devices, and on our side we can update the tables as needed.

File: sound/soc/intel/boards/sof_sdw.c [hunk 3]
```diff
@@ -253,12 +253,21 @@ static struct sof_sdw_codec_info codec_info_list[] = {
 	},
 };
 
-static inline int find_codec_info_part(unsigned int part_id)
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:256]: Shouldn't we modify all the tables to add the version number above? Otherwise you haven't really tested the code?

File: sound/soc/intel/boards/sof_sdw.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/intel/boards/sof_sdw.c]: while we're changing can we also slightly modify the function to return i when the match is found and return -EINVAL outside?

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c]: I don't get your comment @ranj063, we already return i and -EINVAL?

File: sound/soc/intel/boards/sof_sdw.c [hunk 4]
```diff
@@ -253,12 +253,21 @@ static struct sof_sdw_codec_info codec_info_list[] = {
...
 	for (i = 0; i < ARRAY_SIZE(codec_info_list); i++)
-		if (part_id == codec_info_list[i].id)
+		/*
+		 * A codec info is for all sdw version with the part id if
+		 * v_id is not specified in the codec info.
+		 */
+		if (part_id == codec_info_list[i].id &&
+		    (!codec_info_list[i].v_id ||
+		     sdw_version == codec_info_list[i].v_id))
 			break;
```
Comment (lyakh) [inline sound/soc/intel/boards/sof_sdw.c]: in fact you could just `return i;` here and `return -EINVAL;` after the loop unconditionally, maybe this is what @ranj063 meant?

File: sound/soc/intel/boards/sof_sdw.c (hunk 2 above)
Comment (lyakh) [reply sound/soc/intel/boards/sof_sdw.c]: @plbossart that test is exactly for the reason not to have to update all entries of `codec_info_list` immediately. It's a kind of a "backward compatibility" mode - entries without `.v_id` explicitly set will still work, and those, where it's set, since it cannot be 0 as you say, the code will check the version.

File: sound/soc/intel/boards/sof_sdw_common.h (hunk 1 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_common.h]: > we can use 100 characters so let's use part_id and vendor_id. No need to use any encoding.

Fixed

File: sound/soc/intel/boards/sof_sdw.c (hunk 2 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > @plbossart that test is exactly for the reason not to have to update all entries of `codec_info_list` immediately. It's a kind of a "backward compatibility" mode - entries without `.v_id` explicitly set will still work, and those, where it's set, since it cannot be 0 as you say, the code will check the version.

Thanks @lyakh and the other reason is that some codecs with different version id can use the same codec info. For example, we can use the existing  rt715 codec info for rt715 sdca version codec. So a codec info will be used for all codecs with the same part id if version_id is not specified.

File: sound/soc/intel/boards/sof_sdw.c (hunk 4 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > in fact you could just `return i;` here and `return -EINVAL;` after the loop unconditionally, maybe this is what @ranj063 meant?

Fixed

File: sound/soc/intel/boards/sof_sdw.c (hunk 3 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:256]: > Shouldn't we modify all the tables to add the version number above? Otherwise you haven't really tested the code?

Currently, all the existing codec infos are for all version id. I will add the version id for those codec info which is for a specific version only in the follow up patch.

File: sound/soc/intel/boards/sof_sdw.c [hunk 5]
```diff
@@ -253,18 +253,25 @@ static struct sof_sdw_codec_info codec_info_list[] = {
...
 	for (i = 0; i < ARRAY_SIZE(codec_info_list); i++)
-		if (part_id == codec_info_list[i].id)
-			break;
+		/*
+		 * A codec info is for all sdw version with the part id if
+		 * vendor_id is not specified in the codec info.
+		 */
+		if (part_id == codec_info_list[i].id &&
+		    (!codec_info_list[i].vendor_id ||
+		     sdw_version == codec_info_list[i].vendor_id))
```
Comment (lyakh) [inline sound/soc/intel/boards/sof_sdw.c]: it just looks a bit strange for someone, who hasn't read the SDW spec: the "vendor ID" is really the SDW version? Where does this come from? Maybe we should use the same name then or at least add a bit about that to the comment?

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > it just looks a bit strange for someone, who hasn't read the SDW spec: the "vendor ID" is really the SDW version? Where does this come from? Maybe we should use the same name then or at least add a bit about that to the comment?

Thanks for pointing this out. I think vendor_id is a typo. Rename it to version_id now.

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
It's been a while since we saw 'All checks have passed"....
